gensim/models/sent2vec_inner.pyx

```diff
@@ -0,0 +1,445 @@
+    loss = negative_sampling(target, lr, grad, wo, hidden, vector_size, neg, negatives, negpos, negatives_len)
+    sscal(&vector_size, &alpha, grad, &ONE)
+    for i from 0 <= i < context.size():
+        our_saxpy(&vector_size, &ONEF, grad, &ONE, &wi[context[i]], &ONE)
+    return loss
+
+
+cdef REAL_t random_uniform()nogil:
+    """Generate random real number between 0 and 1.
+
+    Returns
+    -------
+    REAT_t
+        Generate random real number between 0 and 1.
+
+    """
+    return rand() / (RAND_MAX + 1.0)
+
+
+cdef int random_range(int a, int b)nogil:
+    """Generate random integer for given input range.
+
+    Parameters
+    ----------
+    a : int
+        Lower limit of input range.
+    b: int
+        Upper limit of input range.
+    Returns
+    -------
+    int
+        Random integer in given input range [a,b].
+
+    """
+    return a + <int>(rand() % ((b - a) + 1))
+
+
+cdef int get_line(vector[int] &wids, vector[int] &words, int max_line_size)nogil:
+    """Converting sentence to a list of word ids inferred from the dictionary.
+
+    Parameters
+    ----------
+    wids : vector[int]
+        Vector of word ids for a sentence.
+    words : vector[int] &
+        Reference to vector of valid word ids for a sentence.
+    max_line_size : int
+        Maximum length of input sentence.
+    Returns
+    -------
+    ntokens : int
+        Number of tokens processed in given sentence.
+
+    """
+    cdef int ntokens = <int> 0
+    cdef int i
+
+    for i from 0 <= i < wids.size():
+        if wids[i] < 0:
+            continue
+        ntokens += 1
+        words.push_back(wids[i])
+        if ntokens >= max_line_size:
+            break
+    return ntokens
+
+
+cdef void add_ngrams_train(vector[int] &line, int n, int k, int bucket, int size)nogil:
+    """Training word ngrams for a given context and target word.
+
+    Parameters
+    ----------
+    line : vector[int] &
+        Reference to vector of word ids for input sentence.
+    n : int
+        Number of word ngrams.
+    k : int
+        Number of word ngrams dropped while training a Sent2Vec model.
+    bucket : int
+        Number of hash buckets for vocabulary.
+    size : int
+        Size of sentence embeddings.
+
+    """
+    cdef int num_discarded = 0
+    cdef vector[int] discard
+    cdef int line_size = line.size()
+    cdef int token_to_discard
+    cdef unsigned int i, j, h
+
+    for i from 0 <= i < line.size():
+        discard.push_back(0)
+
+    while num_discarded < k and line_size - num_discarded > 2:
+        token_to_discard = random_range(0, line_size-1)
+        if discard[token_to_discard] == 0:
+            discard[token_to_discard] = 1
+            num_discarded += 1
+
+    for i from 0 <= i < line_size:
+        if discard[i] == 1:
+            continue
+        h = line[i]
+        for j from i + 1 <= j < line_size:
+            if j >= i + n or discard[j] == 1:
+                break
+            h = h * 116049371 + line[j]
```

Comment:
it's quite not obvious what a magic number is 116049371

Reply:
this is mimic to FB implementation

Reply:
A good idea to include that info in a comment -- so someone doesn't accidentally change the magic numbers in the future.